contracts/utils/cryptography/P256.sol

```diff
@@ -170,11 +171,13 @@ library P256 {
      */
     function _affineFromJacobian(uint256 jx, uint256 jy, uint256 jz) private view returns (uint256 ax, uint256 ay) {
         if (jz == 0) return (0, 0);
-        uint256 zinv = Math.invModPrime(jz, P);
-        uint256 zzinv = mulmod(zinv, zinv, P);
-        uint256 zzzinv = mulmod(zzinv, zinv, P);
-        ax = mulmod(jx, zzinv, P);
-        ay = mulmod(jy, zzzinv, P);
+        uint256 p = P; // cache P on the stack
+        uint256 zinv = Math.invModPrime(jz, p);
+        assembly ("memory-safe") {
+            let zzinv := mulmod(zinv, zinv, p)
+            ax := mulmod(jx, zzinv, p)
+            ay := mulmod(jy, mulmod(zzinv, zinv, p), p)
+        }
```

Comment:
Seems like these 3 variables could be outside of an assembly block. I think we would prefer the vanilla version, the function is called only once per verification so it shouldn't increase gas costs dramatically.

```suggestion
        uint256 zzinv = mulmod(zinv, zinv, p);
        ax = mulmod(jx, zzinv, p);
        ay = mulmod(jy, mulmod(zzinv, zinv, p), p);
```

Reply:
here's the gas report -- i think it's worth it, especially for deployment costs. the difference between inside and outside assembly blocks are almost none:

before:
| test/utils/cryptography/P256.t.sol:GasMeasurer contract |                 |        |        |        |         |
|---------------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |        |        |        |         |
| 639497                                                  | 2741            |        |        |        |         |
| Function Name                                           | min             | avg    | median | max    | # calls |
| verify                                                  | 251708          | 256552 | 256564 | 260836 | 256     |
| verifySolidity                                          | 247735          | 252579 | 252591 | 256863 | 256     |


after:
| test/utils/cryptography/P256.t.sol:GasMeasurer contract |                 |        |        |        |         |
|---------------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |        |        |        |         |
| 623292                                                  | 2666            |        |        |        |         |
| Function Name                                           | min             | avg    | median | max    | # calls |
| verify                                                  | 251633          | 256598 | 256837 | 260761 | 256     |
| verifySolidity                                          | 247660          | 252625 | 252864 | 256788 | 256

Reply:
I tried both 
```solidity
    function _affineFromJacobian(uint256 jx, uint256 jy, uint256 jz) private view returns (uint256 ax, uint256 ay) {
        if (jz == 0) return (0, 0);
        uint256 p = P; // cache P on the stack
        uint256 zinv = Math.invModPrime(jz, p);
        assembly ("memory-safe") {
            let zzinv := mulmod(zinv, zinv, p)
            ax := mulmod(jx, zzinv, p)
            ay := mulmod(jy, mulmod(zzinv, zinv, p), p)
        }
    }
```

and 
```solidity
    function _affineFromJacobian(uint256 jx, uint256 jy, uint256 jz) private view returns (uint256 ax, uint256 ay) {
        if (jz == 0) return (0, 0);
        uint256 p = P; // cache P on the stack
        uint256 zinv = Math.invModPrime(jz, p);
        uint256 zzinv = mulmod(zinv, zinv, p);
        ax = mulmod(jx, zzinv, p);
        ay = mulmod(jy, mulmod(zzinv, zinv, p), p);
    }
```

The assembly version is very very slightly less expensive. really not that significant.